source/base/function_spherical.cc

```diff
@@ -0,0 +1,179 @@
+  H : factor(grads(grads(f)));
+  subst([diff(ur,theta)=sin(phi)*utheta,
+       diff(utheta,theta)=-cos(phi)*uphi-sin(phi)*ur,
+       diff(uphi,theta)=cos(phi)*utheta,
+       diff(ur,phi)=uphi,
+       diff(uphi,phi)=-ur],
+      H);
+
+  where (used later to do tensor products):
+
+  ur     : [cos(theta)*sin(phi), sin(theta)*sin(phi), cos(phi)];
+  utheta : [-sin(theta), cos(theta), 0];
+  uphi   : [cos(theta)*cos(phi), sin(theta)*cos(phi), -sin(phi)];
+
+   */
+
+  template <int dim>
+  Spherical<dim>::Spherical(const Point<dim> &p)
+    :
+    Function<dim>(1),
+    coordinate_system_offset(p)
+  {
+  }
+
+  template <int dim>
+  Spherical<dim>::~Spherical()
+  {
+  }
+
+
+  template <int dim>
+  double
+  Spherical<dim>::value (const Point<dim>   &p_,
+                         const unsigned int) const
+  {
+    AssertThrow(dim==3,
+                ExcNotImplemented());
+
+    const Point<dim> p = p_ - coordinate_system_offset;
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    return svalue(sp);
+  }
+
+
+  template <int dim>
+  void
+  Spherical<dim>::set_unit_vectors(Tensor<1,dim> &unit_r,
+                                   Tensor<1,dim> &unit_theta,
+                                   Tensor<1,dim> &unit_phi,
+                                   const double &cos_theta,
+                                   const double &sin_theta,
+                                   const double &cos_phi,
+                                   const double &sin_phi) const
+  {
+    unit_r[0]     = cos_theta * sin_phi;
+    unit_r[1]     = sin_theta * sin_phi;
+    unit_r[2]     = cos_phi;
+
+    unit_theta[0] = -sin_theta;
+    unit_theta[1] =  cos_theta;
+    unit_theta[2] = 0.;
+
+    unit_phi[0]   = cos_theta * cos_phi;
+    unit_phi[1]   = sin_theta * cos_phi;
+    unit_phi[2]   = -sin_phi;
+  }
+
+
+  template <int dim>
+  Tensor<1,dim>
+  Spherical<dim>::gradient (const Point<dim>   &p_,
+                            const unsigned int) const
+  {
+    AssertThrow(dim==3,
+                ExcNotImplemented());
+
+    const Point<dim> p = p_ - coordinate_system_offset;
+    const std_cxx11::array<double, dim> sp = GeometricUtilities::Coordinates::to_spherical(p);
+    const Tensor<1,dim> sg = sgradient(sp);
+
+    // FIXME: somewhat backwards, but we need cos/sin's for unit vectors
+    const double cos_theta = std::cos(sp[1]);
+    const double sin_theta = std::sin(sp[1]);
+    const double cos_phi   = std::cos(sp[2]);
+    const double sin_phi   = std::sin(sp[2]);
+
+    Tensor<1,dim> unit_r, unit_theta, unit_phi;
+    set_unit_vectors(unit_r, unit_theta, unit_phi,
+                     cos_theta, sin_theta,
+                     cos_phi, sin_phi);
+
+    Tensor<1,dim> res;
+
+    if (sg[0] != 0.)
+      {
+        res += unit_r * sg[0];
+      }
+
+    if (sg[1] != 0.)
+      {
+        Assert (sp[0] > 0.,
+                ExcDivideByZero());
```

Comment:
But `sp` is a `std_cxx11::array<double, dim>`, so `sp[0]` **could** be negative (although one would expect that `GeometricUtilities::Coordinates::to_spherical()` would check that this is not the case).